client/app/scripts/superdesk/notification/notification.js

```diff
@@ -30,19 +34,67 @@
             return;
         }
 
-        var ws = new WebSocket(config.server.ws);
-
-        ws.onmessage = function(event) {
-            var msg = angular.fromJson(event.data);
-            $rootScope.$broadcast(msg.event, msg.extra);
-            if (_.contains(ReloadEvents, msg.event)) {
-                $rootScope.$broadcast('reload', msg);
-            }
+        var connect = function() {
+            ws = new WebSocket(config.server.ws);
+            bindEvents();
         };
 
-        ws.onerror = function(event) {
-            console.error(event);
+        var bindEvents = function() {
+            ws.onmessage = function(event) {
+                var msg = angular.fromJson(event.data);
+                $rootScope.$broadcast(msg.event, msg.extra);
+                if (_.contains(ReloadEvents, msg.event)) {
+                    $rootScope.$broadcast('reload', msg);
+                }
+            };
+
+            ws.onerror = function(event) {
+                console.error('in error ws.onerror');
+                console.error(event);
+            };
+
+            ws.onopen = function(event) {
+                clearInterval(connectTimer);
+                $rootScope.$broadcast('connected');
+            };
+
+            ws.onclose = function(event) {
+                $rootScope.$broadcast('disconnected');
+                clearInterval(connectTimer);
+                connectTimer = setInterval(function() {
+                    if (ws) {
+                        connect();  // Retry to connect for every TIMEOUT interval.
```

Comment:
It it probably better to incrementally increase the timeout period. For example, if a disconnect happens due to a temporarily overloaded server, hammering it with reconnection attempts would do no good, especially if _all_ clients would be trying to do the same.

If reconnection has failed several times in a row already, a client can probably assume that something more serious is going on and the issues are not just momentary, hence less likely to be resolved in the very next attempt. It thus makes sense to wait for a bit longer. When a connection is finally restored, the timeout period can be again reset to the initial short value.

What do you guys think, should we add this smarter reconnecting logic?